### Natural Computing Assignment 3 
Simge Ekiz(s4706757), Luca Parola(s1009497), Katrin Bujari(s1005213)

### Question 1


Consider an illustrative example of a PSO system composed of three particles. Consider the following update rule for each particle i and dimension d:
\begin{equation}
		v(i;d) = wv(i;d) + r_1({x}^*(i;d) - x(i;d)) + r_2({x}^*(d) - x(i;d)) 
\end{equation}


To facilitate calculation, we will ignore the fact that $r_1$ and $r_2$ are random numbers and fix them to $0.5$ for this exercise. The space of solutions is the two dimensional real valued space and the current state of the swarm is as follows

-  Position of particles: $x_1 = (5,5); x_2 = (8,3); x_3 = (6,7)$
-  Individual best positions: $x^*_1 = (5,5); x^*_2 = (7,3); x^*_3 = (5,6)$
-  Social best position: $x^* = (5,5)$
-  Velocities: $v_1 = (2,2); v_2 = (3,3); v_3 = (4,4)$


a-) What would be the next position of each particle after one iteration of the PSO algorithm with w = 2?



<div style="padding-left: 20px;">
$w = 2$
$r_1 = 0.5$
$r_2 = 0.5$
<br><br>
<b> According to formula; </b><br>
  <span style="padding-left: 15px;" >  $v(i;d) = wv(i;d) + r_1(x^∗(i;d) - x(i;d)) + r2(x^∗(d) - x(i;d))$
  </span>
<br><br>
<b> Velocity of particle 1; </b><br> 
<span style="padding-left: 15px;" >
    $v(1;xaxis) = 2*2 + 0.5*(5-5) + 0.5*(5-5) = 4$ <br>
</span>
<span style="padding-left: 15px;" >
    $v(1;yaxis) = 2*2 + 0.5*(5-5) + 0.5*(5-5) = 4$ <br> 
</span>
<span style="padding-left: 15px;" >
    ${v}_1^2 = (4, 4)$ 
</span>
<br><br>
<b> Velocity of particle 2;</b> <br>
<span style="padding-left: 15px;" >
    $v(2;xaxis) = 2*3 + 0.5*(7-8) + 0.5*(5-8) = 4$ <br>
</span>
<span style="padding-left: 15px;" >
    $v(2;yaxis) = 2*3 + 0.5*(3-3) + 0.5*(5-3) = 7$ <br>
</span>
<span style="padding-left: 15px;" >
    $v_2^2 = (4, 7)$ 
</span>
<br><br>
<b> Velocity of particle 3; </b><br>
<span style="padding-left: 15px;" >
    $v(3;xaxis) = 2*4 + 0.5*(5-6) + 0.5*(5-6) = 7$ <br>
</span>
<span style="padding-left: 15px;" >
    $v(3;yaxis) = 2*4 + 0.5*(6-7) + 0.5*(5-7) = 6.5$ <br>
</span>
<span style="padding-left: 15px;" >
    $v_3^2 = (7, 6.5)$ 
</span>
<br><br>
<b> Updating positions : </b> <br>
<span style="padding-left: 15px;" >
    $x_i^{t+1} = x_i^{t} + v_i^{t+1}$
</span><br><br>
<b> Position of particle 1; </b> <br>
<span style="padding-left: 15px;" >
$x_1^2 = (5, 5) + v_1^2 = (5, 5) + (4, 4) = (9, 9) $
</span>
<br><br>
<b> Position of particle 2; </b><br>
<span style="padding-left: 15px; " >
$x_2^2 = (8,3) + v_2^2 = (8, 3) + (4, 7) = (12, 10) $ 
</span> <br><br>
<b>Position of particle 3; </b> <br>
<span style="padding-left: 15px;" >
$x_3^2 = (6,7) + v_3^2 = (6, 7) + (7, 6.5) = (13, 13.5) $ 
</span>
</div>

b-) And using w = 0.1?

<div style="padding: 20px;">
$w=0.1$
$r_1 = 0.5$
$r_2 = 0.5$
<br><br>
<b>Velocity of particle 1;</b> <br> 
<span style="padding-left: 15px;" >
    $v(1;xaxis) = 0.1*2 + 0.5*(5-5) + 0.5*(5-5) = 0.2$ <br>
</span>
<span style="padding-left: 15px;" >
    $v(1;yaxis) = 0.1*2 + 0.5*(5-5) + 0.5*(5-5) = 0.2$ <br> 
</span>
<br>
<b>Velocity of particle 2;</b> <br>
<span style="padding-left: 15px;" >
    $v(2;xaxis) = 0.1*3 + 0.5*(7-8) + 0.5*(5-8) = -1.7$ <br>
</span>
<span style="padding-left: 15px;" >
    $v(2;yaxis) = 0.1*3 + 0.5*(3-3) + 0.5*(5-3) = 1.3$ <br>
</span>
<br>
<b>Velocity of particle 3;</b> <br>
<span style="padding-left: 15px;" >
    $v(3;xaxis) = 0.1*4 + 0.5*(5-6) + 0.5*(5-6) = -0.6$ <br>
</span>
<span style="padding-left: 15px;" >
    $v(3;yaxis) = 0.1*4 + 0.5*(6-7) + 0.5*(5-7) = -1.1$ <br>
</span>
<br><br>

<b>Position of particle 1;</b> <br>
<span style="padding-left: 15px;" >
$x_1^2 = (5, 5) + (0.2, 0.2) = (5.2, 5.2) $
</span>
<br><br>
<b>Position of particle 2;</b> <br>
<span style="padding-left: 15px; " >
$x_2^2 = (8, 3) + (-1.7, 1.3) = (6.3, 4.3) $ 
</span> <br><br>
<b>Position of particle 3; </b> <br>
<span style="padding-left: 15px;" >
$x_3^2 = (6, 7) + (-0.6, -1.1) = (5.4, 5.9) $ 
</span>
</div>

c-) Explain what is the effect of the parameter w.
> In PSO, for each particle, velocity is calculated according to its previous velocity, the particle location which has the best fitness so far and any particles location which has the best fitness so far. W carries a factor that how much the previous velocity contribute to the next velocity. small w facilitates local search, it exploits the current position as much as it can. Large w facilitates global search, it aims to explore solution space.


d-) Give an advantage and a disadvantage of a high value of w.
> An advantage of high value w is that it allows us to explore more. Particles move faster and more solution space could be searced and this increase the possibility of finding a better solution.
A disadvantage of of high value w is that;
it does less expoitation and gets less sensitive to new information in the influence from the current particle. Thus, it might skip over an optimum.





### Question 2

Consider a particle swarm consisting of a single member. How would it perform in a
trivial task such as the minimization of $f(x) = x^2$ when $w < 1$?

> Given that w<1 and $f(x) = x^2$ It would find the optimum solution. Finding a solution depends on the objective function since, f(x) establishes a convex solution space which has less complexity of the search space. For instance, If the particle starts with the velocity which drawns the particle away from the optimum, personal best effect the particle and after some time particle will be moving to optimum point. When velocity is in the same direction with the optimum, personal best loose its influence. Then, velocity start to decrease. If the particle reaches to optimum before the velocity is zero, algorithm finds the optimum point. Particle will find the optimum for this trivial problem, because when w < 1, particles would be moving around the optimum. Then, after some steps velocity would be zero at the optimum.


### Question 3

Implement the PSO algorithm for clustering described in “Van der Merwe, D. W., and Andries Petrus Engelbrecht. ”Data clustering using particle swarm optimization.” Evolutionary Computation, 2003. CEC’03. The 2003 Congress on. Vol. 1. IEEE, 2003.” (see also swarm intelligence slides). Implement the k-means clustering.
Apply and compare the performance of the two algorithms in terms of quantization error on Artificial dataset 1 and Iris dataset (the latter available at UCI ML repository, see https://archive.ics.uci.edu/ml/datasets/iris). In both algorithms, use the true number of clusters as value of the parameter for setting the number of clusters.

In [7]:
from sklearn import datasets
from sklearn.cluster import KMeans
import pandas as pd
import numpy as np
import scipy

We build the Artificial dataset defined as the equation 9 in Van der Merwe, D. W., and
Andries Petrus Engelbrecht. "Data clustering using particle swarm optimization."

In [8]:
z1=np.random.uniform(-1,1,400)
z2=np.random.uniform(-1,1,400)

In [9]:
artificial_df=pd.DataFrame({"Z1" : z1,"Z2" : z2})

In [10]:
kmeans1 = KMeans(n_clusters=2, random_state=0).fit(artificial_df)

We predict the class of each datapoint

In [11]:
prediction_artificial_df=kmeans1.predict(artificial_df)

Now we can compute the quantization error for the artificial dataset

### quantization error

$$J_{e}\,=\,{{\sum_{j=1}^{N_{c}}[\sum_{\forall {\bf z}_{{\rm p}}\in C_{ij}} d({\bf z}_{p},{\bf m}_{j})/\vert C_{ij}\vert]}\over{N_{c}}}$$


$$d ( {{\bf z}_{p}, m_{j}})\,=\,\sqrt{\sum_{k=1}^{N_{j}}(z_{pk}-m_{jk})^{2}}$$



$N_{c}$ denotes the number of cluster centroids (as provided by the user), i.e. the number of dusters to be formed

${\bf z}_{p}$ denotes the p-th data vector

$C_{j}$ is the subset of data vectors that form cluster j.

$m_{j}$ is the centroid 

$d_{{\bf z}_{p}, m_{j}}$ is the Euclidean distance 


In [13]:
##we take the centroids:
centroid1_A_df=kmeans1.cluster_centers_[0]
centroid2_A_df=kmeans1.cluster_centers_[1]

In [14]:
## we assign each data point to the predicted cluster 
cluster1_A=[]
cluster2_A=[]
for i in range(0,len(artificial_df)):
    if prediction_artificial_df[i]==0:
        
        cluster1_A.append(artificial_df.loc[i])
        
    if prediction_artificial_df[i]==1:
        cluster2_A.append(artificial_df.loc[i])

In [15]:
def quantization_error2(cluster1,cluster2,centroid1,centroid2):
            
    distance1=0
    distance2=0


    for i in range(0,len(cluster1)):

        dist1=scipy.spatial.distance.euclidean(np.asarray(cluster1[i]), centroid1)
        distance1=distance1+dist1
    
    if len(cluster1)>0:
        distance1=distance1/len(cluster1)
    else:
        distance1=0

    for i in range(0,len(cluster2)):

        dist2=scipy.spatial.distance.euclidean(np.asarray(cluster2[i]), centroid2)
        distance2=distance2+dist2
    
    if len(cluster2)>0:
        distance2=distance2/len(cluster2)
    else:
        distance2=0


    QE=(distance1+distance2)/2
    ##print("Quantization error: ",QE)
    return QE

In [16]:
print ("The quantization error for the KMEANS algorithm for the artificial dataset is  ", quantization_error2(cluster1_A,cluster2_A,centroid1_A_df,centroid2_A_df))

The quantization error for the KMEANS algorithm for the artificial dataset is   0.5778295493707196


Now we can do the same with the iris dataset

In [17]:
iris = datasets.load_iris()
y=iris.target
iris_df = pd.DataFrame(iris.data, columns=iris.feature_names)

In [18]:
kmeans2 = KMeans(n_clusters=3, random_state=0).fit(iris_df)

In [19]:
prediction_iris_df=kmeans2.predict(iris_df)

In [20]:
##we take the centroids:
centroid1_iris=kmeans2.cluster_centers_[0]
centroid2_iris=kmeans2.cluster_centers_[1]
centroid3_iris=kmeans2.cluster_centers_[2]


In [21]:
## we assign each data point to the predicted cluster 
cluster1_iris=[]
cluster2_iris=[]
cluster3_iris=[]
for i in range(0,len(iris_df)):
    if prediction_iris_df[i]==0:
        cluster1_iris.append(iris_df.loc[i])
    if prediction_iris_df[i]==1:
        cluster2_iris.append(iris_df.loc[i])
    if prediction_iris_df[i]==2:
        cluster3_iris.append(iris_df.loc[i])

We define our qunatization error function for 3 cluster

In [23]:
def quantization_error3(cluster1,cluster2,cluster3,centroid1,centroid2,centroid3):
            
    distance1=0
    distance2=0
    distance3=0
    count=0
    for i in range(0,len(cluster1)):

        dist1=scipy.spatial.distance.euclidean(np.asarray(cluster1[i]), centroid1)
        distance1=distance1+dist1
    
    if len(cluster1)>0:
        distance1=distance1/len(cluster1)
    else:
        distance1=0
    
   
    

    for i in range(0,len(cluster2)):

        dist2=scipy.spatial.distance.euclidean(np.asarray(cluster2[i]), centroid2)
        distance2=distance2+dist2
    
    if len(cluster2)>0:
        distance2=distance2/len(cluster2)
    else:
        distance2=0
    
  
    
    
    for i in range(0,len(cluster3)):

        dist3=scipy.spatial.distance.euclidean(np.asarray(cluster3[i]), centroid3)
        distance3=distance3+dist3
    
    if len(cluster3)>0:
        distance3=distance3/len(cluster3)
    else:
        distance3=0


    
    QE=(distance1+distance2+distance3)/3
    return QE

In [24]:
print ("The quantization error for the KMEANS algorithm for the iris dataset is  ", quantization_error3(cluster1_iris,cluster2_iris,cluster3_iris,centroid1_iris,centroid2_iris,centroid3_iris))

The quantization error for the KMEANS algorithm for the iris dataset is   0.6473743892616001


Now we can build the PSO algorithm and use it with the Artificial DATaset

In [28]:
#we initialize 10 particles with 2 centroid each to use in the Artificial dataset
particles_list=[]
for i in range(0,10):
    particles=np.random.uniform(-1,1,4).reshape(2,2)
    particles_list.append(particles)

In [29]:
##we set our initial values

personal_best_error=np.ones(10)
global_best_position=[]
global_best_error=1
personal_best_position=np.zeros(40).reshape(10,2,2)

alpha=1.49618
w=0.7298
velocity=np.zeros(4).reshape(2,2) 

##for each iteration
for iteration in range(0,10):
    
    quantization_error=[] 
    count=0
    
    ##for each particles 
    for i in range(0,len(particles_list)):

        
        current_particle=particles_list[i]
        cluster1=[]
        cluster2=[]
        
       ##for each data point in the Artificial dataset
        for a in range(0,len(artificial_df)):
            
            
            ##we measure the euclidean distance from each centroid
            dist1=scipy.spatial.distance.euclidean(np.asarray(artificial_df.loc[a]), current_particle[0])
            dist2=scipy.spatial.distance.euclidean(np.asarray(artificial_df.loc[a]), current_particle[1])

            ##we assign each datapoint to a cluster based on the closest centroid
            if dist1<dist2:
                cluster1.append(artificial_df.loc[a])


            elif dist1>dist2:
                cluster2.append(artificial_df.loc[a])


        
        ##Once we have our cluster we compute the quantization error and we store it
        ## we will have a list with all our quantization error
        quantization_error.append(quantization_error2(cluster1,cluster2,current_particle[0],current_particle[1]))

   
    
    ##Then for each quantization error in the list
    for z in range(0,len(quantization_error)):
        count=count+1    
        
        ##if the quantization error of the current iteration is smaller than the personal best , we store it as new personal best
        if quantization_error[z]<personal_best_error[z]:
            personal_best_error[z]=quantization_error[z]
            ##and we store the corresponding particle as new best position
            personal_best_position[z]=particles_list[z]

        ##we also update the best global error if we find a new best 
        if personal_best_error[z]<global_best_error:
            global_best_error=personal_best_error[z]
            global_best_position=particles_list[z]
            
    ###print ("Particle list BEFORE, ", particles_list)
    
    
    
    
    ## then for each particle we update the position of the centroids based on the formula
    
    for i in range(0,len(particles_list)):
        
        ##I've disassembled the formula for clarity.

        

        ##we set the random r
        r1=np.random.uniform(0,1,1)
        r2=np.random.uniform(0,1,1)

        ##we compute the first term of the formula
        first_term=np.multiply(w,velocity)
        
        ##we compute the two multiplication of the alpha and the r
        alphaR1=alpha*r1
        alphaR2=alpha*r2
        
        #we copmute the 2 subtractions
        first_subtraction=np.subtract(personal_best_position[i],particles_list[i])
        second_subtraction=np.subtract(global_best_position,particles_list[i])

        #we compute the second term
        second_term=np.multiply(alphaR1,first_subtraction)
        
        #we compute the second term
        third_term=np.multiply(alphaR2,second_subtraction)
        
        ##we compute the velocity
        velocity=np.add(first_term,second_term)
        velocity=np.add(velocity,third_term)

        #we update the position of the centroids of the particles
        particles_list[i]=np.add(particles_list[i],velocity)
        
    ##print ("global best error is: ", global_best_error)
    ##print ("global best position: ", global_best_position)
    ##print ("personal best error: ", personal_best_error)
    ##print ("Particle list AFTER", particles_list)
    ##print("personal best position  ",personal_best_position )

    
print("After ",iteration," iteration , the global best error is: ", global_best_error, "best  swarn has positions: ", global_best_position)

After  9  iteration , the global best error is:  0.4570523934830559 best  swarn has positions:  [[ 0.53278471 -0.25095812]
 [ 1.87142303 -1.13583857]]


Now we can di the same thing for the irisi dataset. This time as we have 3 classes and 4 dimensions we set each particle to have 3 centroids based on 4 dmensions

We take the max and min of each columns in the iris dataset, so we can set limits to our randomly chosen centroids

In [35]:
max_sepal_length=max(iris_df.iloc[:,0])
min_sepal_length=min(iris_df.iloc[:,0])

max_sepal_width=max(iris_df.iloc[:,1])
min_sepal_width=min(iris_df.iloc[:,1])

max_petal_length=max(iris_df.iloc[:,2])
min_petal_length=min(iris_df.iloc[:,2])

max_petal_width=max(iris_df.iloc[:,3])
min_petal_width=min(iris_df.iloc[:,3])

In [38]:
#we initialize 10 particles with 3 centroid each and 4 dimensions to use in the iris dataset
particles_list_iris=[]

for i in range(0,10):
    particle=[]
    
    for i in range(0,3):
        centroid=[np.random.uniform(min_sepal_length,max_sepal_length,1),np.random.uniform(min_sepal_width,max_sepal_width,1),np.random.uniform(min_petal_width,max_petal_width,1),np.random.uniform(min_petal_width,max_petal_width,1)]
        particle.append(centroid)
    
    particles_list_iris.append(particle)
    
particles_list_iris=np.asarray(particles_list_iris).reshape(10,3,4)

Now we can run the algorithm

In [39]:
##we set our initial values

personal_best_error=np.full((1, 10), 50).astype(float)
personal_best_error=personal_best_error[0][:]
global_best_position=[]
global_best_error=100  ##we set this to 100 so we are sure that the algorithm will find a smaller value
personal_best_position=np.zeros(120).reshape(10,3,4)

alpha=1.49618
w=0.7298
velocity=np.zeros(12).reshape(3,4) 

##for each iteration
for iteration in range(0,20):
    
    quantization_error=[] 
    count=0
    
    ##for each particles 
    for i in range(0,len(particles_list_iris)):

        
        current_particle=particles_list_iris[i]
        cluster1=[]
        cluster2=[]
        cluster3=[]
        
       ##for each data point in the Artificial dataset
        for a in range(0,len(iris_df)):
            
            
            ##we measure the euclidean distance from each centroid
            dist1=scipy.spatial.distance.euclidean(np.asarray(iris_df.loc[a]), current_particle[0])
            dist2=scipy.spatial.distance.euclidean(np.asarray(iris_df.loc[a]), current_particle[1])
            dist3=scipy.spatial.distance.euclidean(np.asarray(iris_df.loc[a]), current_particle[2])
            
            
            ##we assign each datapoint to a cluster based on the closest centroid
            if dist1<dist2 and dist1<dist3:
                cluster1.append(iris_df.loc[a])


            elif dist2<dist3 and dist2<dist3:
                cluster2.append(iris_df.loc[a])
                
            elif dist3<dist1 and dist3<dist1:
                cluster3.append(iris_df.loc[a])


        
        ##Once we have our cluster we compute the quantization error and we store it
        ## we will have a list with all our quantization error
        quantization_error.append(quantization_error3(cluster1,cluster2,cluster3,current_particle[0],current_particle[1],current_particle[2]))

   
    
    ##Then for each quantization error in the list
    for z in range(0,len(quantization_error)):
        count=count+1    
        
        ##if the quantization error of the current iteration is smaller than the personal best , we store it as new personal best
        if quantization_error[z]<personal_best_error[z]:
            personal_best_error[z]=quantization_error[z]
            ##and we store the corresponding particle as new best position
            personal_best_position[z]=particles_list_iris[z]

        ##we also update the best global error if we find a new best 
        if personal_best_error[z]<global_best_error:
            global_best_error=personal_best_error[z]
            global_best_position=particles_list_iris[z]
            
    ###print ("Particle list BEFORE, ", particles_list)
    
    
    
    
    ## then for each particle we update the position of the centroids based on the formula
    
    for i in range(0,len(particles_list_iris)):
        
        ##I've disassembled the formula for clarity.

        

        ##we set the random r
        r1=np.random.uniform(0,1,1)
        r2=np.random.uniform(0,1,1)

        ##we compute the first term of the formula
        first_term=np.multiply(w,velocity)
        
        ##we compute the two multiplication of the alpha and the r
        alphaR1=alpha*r1
        alphaR2=alpha*r2
        
        #we copmute the 2 subtractions
        first_subtraction=np.subtract(personal_best_position[i],particles_list_iris[i])
        second_subtraction=np.subtract(global_best_position,particles_list_iris[i])

        #we compute the second term
        second_term=np.multiply(alphaR1,first_subtraction)
        
        #we compute the second term
        third_term=np.multiply(alphaR2,second_subtraction)
        
        ##we compute the velocity
        velocity=np.add(first_term,second_term)
        velocity=np.add(velocity,third_term)

        #we update the position of the centroids of the particles
        particles_list_iris[i]=np.add(particles_list_iris[i],velocity)
        
    ##print ("global best error is: ", global_best_error)
    ##print ("global best position: ", global_best_position)
    ##print ("personal best error: ", personal_best_error)
    ##print ("Particle list AFTER", particles_list)
    ##print("personal best position  ",personal_best_position )

    
print("After ",iteration," iteration , the global best error is: ", global_best_error, "best  swarn has positions: ", global_best_position)

After  19  iteration , the global best error is:  0.8999702242159057 best  swarn has positions:  [[ 4.25079710e+07 -1.31076707e+07 -1.39633037e+08  1.84153135e+07]
 [-1.10867177e+08  6.51295081e+07  5.39108174e+07 -3.28589939e+07]
 [ 7.82919159e+07 -1.23255213e+07 -1.11076469e+08  2.35742279e+06]]


If we compare the results with the results described in the paper, we can notice that our PSO and KMEANS algorithm performs somehow similarly to the results of the paper. In fact we have:

- Quantization error for the KMEans algorithm for the articifial dataset:   0.5778295493707196
- Quantization error for the KMEans algorithm for the iris dataset:   0.6473743892616001

- Quantization error for the PSO algorithm for the articifial dataset:  0.4570523934830559 
- Quantization error for the PSO algorithm for the articifial dataset::   0.8999702242159057 

#### Question 4

The figure shows an example from the ACO book by Dorigo and Stuetzle. What results do you expect for an ant colony algorithm that does not use tabu lists (except for inhibition of immediate return to the previous node)?

<img src="ACO.png" alt="Drawing" style="width: 30%;"/>

> ACO is inspired from the foraging behaviour of ants. How they find the closest food sources to their nests. They communicate with pheromones they leave. They leave their pheromone while moving. Also they tend to choose paths marked by strong pheromone concentrations. When ant finds a food source, it leaves pheromones according to some probability that may depend on the quantity and quality of the food during the return trip. Thus, quality food sources closest to the nest will have the path with more pheromone. 
However in this case without a tabu list, it is unavoidable that ants make cycles. Bottom of the picture, there are a lot of connected paths and this will cause a lot of pheromone accumulation and this result with most likely loopy paths. For the graph like this converging the minumum path is not trivial.
To find the optimum solution an ant has to make number of correct desicions otherwise ant generates suboptimal solutions.
There is a trade off between use of an easy but suboptimal path and searching the optimal path. 
In this case easy path would be that the ants takes the route trough the upper nodes and there is two optimal paths which are in the lower part of the graph and harder to find.
We can avoid from cycles with high pheromone evaporation However, this will end up with not generating optimum solutions. Since this problem is relatively simple problem because of the large number of ants compared to the relatively small search space. After a lot of iterations algorithm find the fastest route in the densely connected part of graph which is down-up-right-down-up or down-down-rigt-up-up.


#### Question 5

Assume that ants are allowed to lay pheromone on a path at every time step, so that the pheromone update rule is applied at each time step. Come up with a combination local/global updating scheme that encourages exploration and exploitation - consider which parameters influence this.

> Ant is probabilisticaly choose the next node based on the amount of pheromone on the next node while it is searching for food exploring the space.
For each ant we can keep a limited memory which is local pheromone this allows them to record their trip to their destination. When they reach their destination they can construct a way to back home. If they've passed over the same node twice they can eliminate the path in between those repeat nodes. Then they will retrace the steps they constructed based on the memory(local pheromone). On the way back to the nest ant will ignore the pheromone trails, follow the path that is costructed.
When they are going back to the nest they will be applying pheromones to the ground(global pheromones) they will not apply to the global pheromones on the food searching step because that can cause loops.
After each round the new pheromones are added to the old ones. For better exploitation, good ants should influence the new pheromones more than others.
